src/socketio.quiver.js

```diff
@@ -966,6 +966,56 @@ QuiverSocialProvider.prototype.addFriend_ = function(servers, userId, pubKey,
   return this.syncConfiguration_().then(function() {
     // TODO: Connect to all servers.
     return this.connectAsClient_(friendDesc, servers[0]).ready;
+  }.bind(this)).catch(function(err) {
+    // The connection to the invitation server has failed.  This could be
+    // because the front domain is malfunctioning.  Try again with one of the
+    // default front domains.  This last-ditch effort is worthwhile because
+    // (1) invitations only contain a single server, so the front domain is a
+    // single point of failure, and (2) there is a known related bug, #2619.
+    var server = servers[0];
+    if (!server.front) {
+      // This isn't a fronted server, so there's nothing to retry.
+      throw err;
+    }
+    var domainComponents = server.domain.split('.');
+    if (domainComponents.length < 3) {
+      // For very short domains, we don't know how to identify which CDN they
+      // use.
+      throw err;
+    }
+    var cloudDomain = domainComponents.slice(1).join('.');
+    var alternateFront = null;
+    // Find the first front domain for this cloud in the default set
+    for (var i = 0; i < QuiverSocialProvider.DEFAULT_SERVERS_.length; ++i) {
+      var defaultServer = QuiverSocialProvider.DEFAULT_SERVERS_[i];
+      if (defaultServer.domain.endsWith(cloudDomain) &&
+          defaultServer.type === server.type &&
+          defaultServer.front) {
+        alternateFront = defaultServer.front;
+        break;
+      }
+    }
+    if (!alternateFront) {
+      // We didn't find any default server on the same cloud, so we don't know
+      // of any alternate front domain.
+      throw err;
+    }
+    if (alternateFront === server.front) {
+      // The front we found was the same one we started with.  This will happen
+      // whenever the connection failure is due to a non-front-related network
+      // problem.  The deterministic ordering of DEFAULT_SERVERS, combined with
+      // this check, avoids an unbounded retry loop.
+      throw err;
```

Comment:
We might also want to add a comment like:
```
CONSIDER: Retrying with each of the possible front domains, keeping track of the ones tried already to avoid infinite recursion.
```